### PR TITLE
feat: position actions

### PR DESCRIPTION
**Pull request by ChefJerry**
---

## PR-Codex overview
This PR focuses on restructuring hooks and components related to account positions in universal farms. It adds new hooks, updates existing ones, and refactors components for better functionality.

### Detailed summary
- Deleted `hooks.ts` related to account positions
- Added new hooks for account positions
- Updated components for position items and actions
- Refactored stake modal components and utils
- Added new fetcher functions for V3 farming rewards

> The following files were skipped due to too many changes: `apps/web/src/state/farmsV4/state/accountPositions/fetcher/v3/getAccountV3FarmingPendingCakeReward.ts`, `apps/web/src/state/farmsV4/state/accountPositions/type.ts`, `apps/web/src/state/farmsV4/state/accountPositions/hooks/useAccountStableLpDetails.ts`, `apps/web/src/state/farmsV4/state/accountPositions/fetcher/v3/getAccountV3TokenIds.ts`, `apps/web/src/views/universalFarms/components/PositionItem/PriceRange.tsx`, `apps/web/src/views/PoolDetail/components/PoolEarnings.tsx`, `apps/web/src/views/universalFarms/hooks/useV2FarmActions.ts`, `apps/web/src/views/universalFarms/hooks/useCakeEarning.ts`, `apps/web/src/state/farmsV4/state/accountPositions/fetcher/v3/getAccountV3Positions.ts`, `apps/web/src/state/farmsV4/state/accountPositions/hooks/useAccountV2LpDetails.ts`, `apps/web/src/state/farmsV4/state/accountPositions/hooks/useExtraV3PositionInfo.ts`, `apps/web/src/views/universalFarms/components/PositionItem/PositionItem.tsx`, `apps/web/src/views/universalFarms/components/PositionItem/V3PositionItem.tsx`, `apps/web/src/views/universalFarms/components/PositionItem/StablePositionItem.tsx`, `apps/web/src/state/farmsV4/state/accountPositions/hooks/useAccountPositionDetailByPool.ts`, `apps/web/src/views/universalFarms/components/PositionActions/StakeActions.tsx`, `apps/web/src/views/universalFarms/components/PositionItem/V2PositionItem.tsx`, `apps/web/src/views/universalFarms/PositionPage.tsx`, `apps/web/src/views/universalFarms/components/PositionItem/PositionInfo.tsx`, `apps/web/src/views/universalFarms/components/PositionActions/V3PositionActions.tsx`, `apps/web/src/views/universalFarms/components/PositionActions/V2PositionActions.tsx`, `apps/web/src/views/PoolDetail/components/MyPositions.tsx`, `apps/web/src/state/farmsV4/state/accountPositions/fetcher.ts`

> ✨ Ask PR-Codex anything about this PR by commenting with `/codex {your question}`

### CODE REVIEW

File: apps/web/src/state/farmsV4/state/accountPositions/fetcher/v3/getAccountV3FarmingPendingCakeReward.ts [hunk 1]
```diff
@@ -0,0 +1,49 @@
+import { ChainId } from '@pancakeswap/chains'
+import { masterChefV3ABI } from '@pancakeswap/v3-sdk'
+import { getMasterChefV3Contract } from 'utils/contractHelpers'
+import { Address, decodeFunctionResult, encodeFunctionData, Hex } from 'viem'
+
+export const getAccountV3FarmingPendingCakeReward = async (
```
Comment (chef-eric) [inline apps/web/src/state/farmsV4/state/accountPositions/fetcher/v3/getAccountV3FarmingPendingCakeReward.ts:6]: there is `useStakedPositionsByUser` existing, why don't just reuse it by pass a chainId

File: apps/web/src/state/farmsV4/state/accountPositions/hooks/useAccountPositionDetailByPool.ts [hunk 2]
```diff
@@ -0,0 +1,76 @@
...
+  chainId: number,
+  account?: Address | null,
+  poolInfo?: PoolInfo,
+): UseQueryResult<PoolPositionDetail[TProtocol]> => {
+  const [currency0, currency1] = useMemo(() => {
+    if (!poolInfo) return [undefined, undefined]
+    const { token0, token1 } = poolInfo
+    return [token0.wrapped, token1.wrapped]
+  }, [poolInfo])
+  const protocol = useMemo(() => poolInfo?.protocol, [poolInfo])
```
Comment (chef-eric) [inline apps/web/src/state/farmsV4/state/accountPositions/hooks/useAccountPositionDetailByPool.ts:27]: I think useMemo here is unnecessary, maybe it just increases memory usage.

File: apps/web/src/state/farmsV4/state/accountPositions/hooks/useAccountPositionDetailByPool.ts [hunk 3]
```diff
@@ -0,0 +1,76 @@
...
+import { Address, isAddressEqual } from 'viem'
+import { PoolInfo } from '../../type'
+import { getAccountV2LpDetails, getStablePairDetails } from '../fetcher'
+import { getAccountV3Positions } from '../fetcher/v3'
+import { PositionDetail, StableLPDetail, V2LPDetail } from '../type'
+
+type PoolPositionDetail = {
+  [Protocol.STABLE]: StableLPDetail
+  [Protocol.V2]: V2LPDetail
+  [Protocol.V3]: PositionDetail[]
```
Comment (chef-eric) [inline apps/web/src/state/farmsV4/state/accountPositions/hooks/useAccountPositionDetailByPool.ts:14]: why v3 type is array?

File: apps/web/src/state/farmsV4/state/accountPositions/hooks/useAccountPositionDetailByPool.ts [hunk 4]
```diff
@@ -0,0 +1,76 @@
...
+      const stablePair = LegacyRouter.stableSwapPairsByChainId[chainId].find((pair) => {
+        return isAddressEqual(pair.stableSwapAddress, poolInfo?.lpAddress as Address)
+      })
+      return getStablePairDetails(chainId, account!, stablePair ? [stablePair] : [])
+    }
+    if (protocol === 'v3') {
+      return getAccountV3Positions(chainId, account!)
+    }
+    return Promise.resolve([])
+  }, [account, chainId, currency0, currency1, poolInfo, protocol])
```
Comment (chef-eric) [inline apps/web/src/state/farmsV4/state/accountPositions/hooks/useAccountPositionDetailByPool.ts:46]: dep of poolInfo can change to poolInfo?.lpAddress to reduce changes, I figure out there are too much changes of object reference in the existing code

File: apps/web/src/state/farmsV4/state/poolApr/fetcher.ts [hunk 5]
```diff
@@ -147,9 +148,18 @@ const calcV3PoolApr = ({
...
 
-export const getUniversalBCakeWrapperForPool = (pool: PoolInfo) => {
+export const getUniversalBCakeWrapperForPool = (pool: { lpAddress: Address; chainId: number; protocol?: Protocol }) => {
+  let { lpAddress } = pool
+  if (pool.protocol === 'stable') {
+    const stablePair = LegacyRouter.stableSwapPairsByChainId[pool.chainId].find((pair) => {
+      return isAddressEqual(pair.stableSwapAddress, pool.lpAddress)
+    })
+    if (stablePair) {
+      lpAddress = stablePair.liquidityToken.address
```
Comment (chef-eric) [inline apps/web/src/state/farmsV4/state/poolApr/fetcher.ts:158]: there are something easy to cause problem if someone don't have context (I means pool.lpAddress of stablePool is not the real lpAddress, it's the stableSwapAddress), i think we can fix the PoolInfo field

File: apps/web/src/state/farmsV4/state/utils.ts [hunk 6]
```diff
@@ -12,8 +13,18 @@ export const parseFarmPools = (
   options: { isFarming?: boolean } = {},
 ): PoolInfo[] => {
   return data.map((pool) => {
+    const localFarm = UNIVERSAL_FARMS.find(
```
Comment (chef-eric) [inline apps/web/src/state/farmsV4/state/utils.ts:16]: the complexity is almost O(N^2), and you know the data and UNIVERSAL_FARMS are both large object, i think there may cause performance issues.

File: apps/web/src/views/PoolDetail/components/MyPositions.tsx [hunk 7]
```diff
@@ -155,10 +177,16 @@ export const MyPositions: React.FC<{ poolInfo: PoolInfo }> = ({ poolInfo }) => {
...
+                setHandleHarvestAll={setHandleHarvestAll}
               />
             ) : null}
             {['v2', 'stable'].includes(poolInfo.protocol) ? (
-              <MyV2OrStablePositions poolInfo={poolInfo} setCount={setCount} setTotalTvlUsd={setTotalLiquidityUSD} />
+              <MyV2OrStablePositions
+                poolInfo={poolInfo}
+                setCount={setCount}
+                setTotalTvlUsd={setTotalLiquidityUSD}
+                setHandleHarvestAll={setHandleHarvestAll}
```
Comment (chef-eric) [inline apps/web/src/views/PoolDetail/components/MyPositions.tsx:188]: better to use a props like `isHarvestAll` to control the child method, avoid the child change back to parent's state

File: apps/web/src/views/universalFarms/hooks/useCakeEarning.ts [hunk 8]
```diff
@@ -0,0 +1,50 @@
...
+    return new BigNumber(earningsAmount ?? 0).times(cakePrice.toString()).toNumber()
+  }, [cakePrice, earningsAmount])
+
+  return {
+    earningsAmount,
+    earningsBusd,
+  }
+}
+
+export const useV3CakeEarning = (tokenIds: bigint[] = []) => {
```
Comment (chef-eric) [inline apps/web/src/views/universalFarms/hooks/useCakeEarning.ts:36]: why the `useV3CakeEarning` not just use pool as arguments, keep same as the `useV2CakeEarning`

File: apps/web/src/state/farmsV4/state/accountPositions/hooks/useAccountPositionDetailByPool.ts (hunk 3 above)
Comment (ChefJerry) [reply apps/web/src/state/farmsV4/state/accountPositions/hooks/useAccountPositionDetailByPool.ts:14]: one V3 pool can have multi position per account

File: apps/web/src/views/universalFarms/hooks/useCakeEarning.ts (hunk 8 above)
Comment (ChefJerry) [reply apps/web/src/views/universalFarms/hooks/useCakeEarning.ts:36]: the only needed arg is tokenIds, why we need the rest?

File: apps/web/src/views/PoolDetail/components/MyPositions.tsx (hunk 7 above)
Comment (ChefJerry) [reply apps/web/src/views/PoolDetail/components/MyPositions.tsx:188]: what `isHarvestAll` for? this is actually setState on purpose

File: apps/web/src/state/farmsV4/state/poolApr/fetcher.ts (hunk 5 above)
Comment (ChefJerry) [reply apps/web/src/state/farmsV4/state/poolApr/fetcher.ts:158]: yes, it should be more clearly, I have stuck with this part many times

File: apps/web/src/state/farmsV4/state/accountPositions/fetcher/v3/getAccountV3FarmingPendingCakeReward.ts (hunk 1 above)
Comment (ChefJerry) [reply apps/web/src/state/farmsV4/state/accountPositions/fetcher/v3/getAccountV3FarmingPendingCakeReward.ts:6]: actually it is a copycat of `useStakedPositionsByUser`, I dont want to modify the legacy code, which will helpful when we do code clean up for dead code

File: apps/web/src/views/PoolDetail/components/MyPositions.tsx (hunk 7 above)
Comment (chef-eric) [reply apps/web/src/views/PoolDetail/components/MyPositions.tsx:188]: the child set a callback(_handleHarvestAll) back to the parent, it's not a best practice.
i means you can pass a `isHarvestAll` to the child, and the child do the `handleHarvestAll` for itself, avoid to setstate back to the parent

**Pull request merged**